packages/flutter/lib/src/material/list_tile.dart

```diff
@@ -588,115 +618,112 @@ class ListTile extends StatelessWidget {
+    if (theme.useMaterial3) {
+      assert(!isDense, 'ListTile.dense cannot be true while Theme.useMaterial3 is true.');
     }
-    final Color? color = _textColor(theme, tileTheme, textStyle.color);
-    return _isDenseLayout(theme, tileTheme)
-      ? textStyle.copyWith(fontSize: 13.0, color: color)
-      : textStyle.copyWith(color: color);
+    return isDense;
   }
 
-  TextStyle _subtitleTextStyle(ThemeData theme, ListTileThemeData tileTheme) {
-    final TextStyle textStyle = theme.useMaterial3 ? theme.textTheme.bodyMedium! : theme.textTheme.bodyMedium!;
-    final Color? color = _textColor(
-      theme,
-      tileTheme,
-      theme.useMaterial3 ? theme.textTheme.bodySmall!.color : theme.textTheme.bodySmall!.color,
-    );
-    return _isDenseLayout(theme, tileTheme)
-      ? textStyle.copyWith(color: color, fontSize: 12.0)
-      : textStyle.copyWith(color: color);
-  }
-
-  TextStyle _trailingAndLeadingTextStyle(ThemeData theme, ListTileThemeData tileTheme) {
-    final TextStyle textStyle = theme.useMaterial3 ? theme.textTheme.bodyMedium! : theme.textTheme.bodyMedium!;
-    final Color? color = _textColor(theme, tileTheme, textStyle.color);
-    return textStyle.copyWith(color: color);
-  }
-
-  Color _tileBackgroundColor(ThemeData theme, ListTileThemeData tileTheme) {
+  // TODO(TahaTesser): Refactor this to support list tile states
+  // similar to `effectiveColor` and `effectiveIconColor` without
+  // other list tile widgets that depend on this implementation.
+  Color _tileBackgroundColor(ThemeData theme, ListTileThemeData tileTheme, ListTileThemeData defaults) {
     final Color? color = selected
       ? selectedTileColor ?? tileTheme.selectedTileColor ?? theme.listTileTheme.selectedTileColor
       : tileColor ?? tileTheme.tileColor ?? theme.listTileTheme.tileColor;
-    return color ?? Colors.transparent;
+    return color ?? defaults.tileColor!;
   }
 
   @override
   Widget build(BuildContext context) {
     assert(debugCheckHasMaterial(context));
     final ThemeData theme = Theme.of(context);
     final ListTileThemeData tileTheme = ListTileTheme.of(context);
-    final IconThemeData iconThemeData = IconThemeData(color: _iconColor(theme, tileTheme));
+    final ListTileStyle listTileStyle = style
+      ?? tileTheme.style
+      ?? theme.listTileTheme.style
+      ?? ListTileStyle.list;
+    final ListTileThemeData defaults = theme.useMaterial3
+        ? _LisTileDefaultsM3(context)
+        : _LisTileDefaultsM2(context, listTileStyle);
+    final Set<MaterialState> states = <MaterialState>{
+      if (!enabled) MaterialState.disabled,
+      if (selected) MaterialState.selected,
+    };
 
-    TextStyle? leadingAndTrailingTextStyle;
+    Color? resolveColor(Color? color) {
+      if (color is MaterialStateColor) {
+        return MaterialStateProperty.resolveAs<Color?>(color, states);
+      }
+      return null;
+    }
+
+    final Color? effectiveIconColor = resolveColor(iconColor)
```

Comment:
This is pretty terrifying, but I suppose that's the price of adding flexibility and maintaining backwards compatibility. I think you could simplify the expressions just a little more by giving _IndividualOverrides an explicitColor parameter and moving the resolveColor expression into _IndividualOverrides:
```dart
final Color? effectiveIconColor =
    _IndividualOverrides(
       explicitColor: iconColor,
       selectedColor: selectedColor,
       enabledColor: iconColor,
     ).resolve(states)
  ?? _IndividualOverrides(
       explicitColor: tileTheme.iconColor,
       selectedColor: tileTheme.selectedColor,
       enabledColor: tileTheme.iconColor,
     ).resolve(states)
  ?? _IndividualOverrides(
       explicitColor: theme.listTileTheme.iconColor,
       selectedColor: theme.listTileTheme.selectedColor,
       enabledColor: theme.listTileTheme.iconColor,
     ).resolve(states)
  ?? _IndividualOverrides(
       selectedColor: defaults.selectedColor,
       enabledColor: defaults.iconColor,
       disabledColor: theme.disabledColor,
     ).resolve(states);
```

Given that, I suppose you could simplify further by making your local resolveColor function be
```dart
Color? resolveColor(Set<MaterialState> states, Color? explicitColor, Color? selectedColor, Color? enabledColor, [Color? disabledColor]) {
  return _IndividualOverrides(
    explicitColor: iconColor,
    selectedColor: selectedColor,
    enabledColor: iconColor,
  ).resolve(states);
}
```

Note also: What we're doing here, using the value of `ListTileTheme.of(context)` _and_ the value of Theme.of(context).listTileTheme isn't typical of the Material components; usually only the first value is used.  The way it's done here is arguably more useful since the defaults from the overall theme and from an optional local theme are effectively merged, however it's complicated to explain and somewhat messier to implement. That said, in this case we're concerned about backwards compatibility, so I think this is OK.

Reply:
> Note also: What we're doing here, using the value of ListTileTheme.of(context) and the value of Theme.of(context).listTileTheme isn't typical of the Material components; usually only the first value is used.
 
I did notice that and I didn't want to remove `Theme.of(context).listTileTheme`, in case it breaks something.